addons/boa/tasks.py

```diff
@@ -0,0 +1,198 @@
+                 f'file_name=[{query_file_name}], full_path=[{file_full_path}], url=[{query_download_url}] ...')
+    download_request = request.Request(query_download_url)
+    download_request.add_header('Cookie', f'{osf_settings.COOKIE_NAME}={cookie_value}')
+    try:
+        boa_query = ensure_str(request.urlopen(download_request).read())
+    except (ValueError, HTTPError, URLError):
+        message = f'Failed to download Boa query file: user=[{user_guid}], project=[{project_guid}], ' \
+                  f'file_name=[{query_file_name}], full_path=[{file_full_path}], url=[{query_download_url}] ...'
+        await sync_to_async(handle_boa_error)(message, BoaErrorCode.UNKNOWN, user.username, user.fullname, project_url,
+                                              file_full_path, query_file_name=query_file_name)
+        return BoaErrorCode.UNKNOWN
+    logger.info('Boa query successfully downloaded.')
+    logger.debug(f'Boa query:\n########\n{boa_query}\n########')
+
+    logger.debug('Boa client opened.')
+    client = BoaClient(endpoint=host)
+    logger.debug(f'Checking Boa credentials: boa_username=[{username}], boa_host=[{host}] ...')
+    try:
+        client.login(username, password)
+    except BoaException:
+        client.close()
+        message = f'Boa login failed: boa_username=[{username}], boa_host=[{host}]!'
+        await sync_to_async(handle_boa_error)(message, BoaErrorCode.AUTHN_ERROR, user.username, user.fullname, project_url,
+                                              file_full_path, query_file_name=query_file_name)
+        return BoaErrorCode.AUTHN_ERROR
+    logger.info('Boa login completed.')
+
+    logger.debug(f'Retrieving Boa dataset: dataset=[{query_dataset}] ...')
+    try:
+        dataset = client.get_dataset(query_dataset)
+    except BoaException:
+        client.close()
+        message = f'Failed to retrieve or verify the target Boa dataset: dataset=[{query_dataset}]!'
+        await sync_to_async(handle_boa_error)(message, BoaErrorCode.UNKNOWN, user.username, user.fullname, project_url,
+                                              file_full_path, query_file_name=query_file_name)
+        return BoaErrorCode.UNKNOWN
+    logger.info('Boa dataset retrieved.')
+
+    logger.debug(f'Submitting the query to Boa API: boa_host=[{host}], dataset=[{query_dataset}] ...')
+    try:
+        boa_job = client.query(boa_query, dataset)
+    except BoaException:
+        client.close()
+        message = f'Failed to submit the query to Boa API: : boa_host=[{host}], dataset=[{query_dataset}]!'
+        await sync_to_async(handle_boa_error)(message, BoaErrorCode.UNKNOWN, user.username, user.fullname, project_url,
+                                              file_full_path, query_file_name=query_file_name)
+        return BoaErrorCode.UNKNOWN
+    logger.info('Query successfully submitted.')
+    logger.debug(f'Waiting for job to finish: job_id=[{str(boa_job.id)}] ...')
+    while boa_job.is_running():
```

Comment:
Is it possible boa jobs might get stuck, causing this to be an infinite loop? Should there be a sleep threshold of, e.g. 24hr, after which we cancel things?

Reply:
ah, it is possible!!! since we don't control the `boa` job. We need to add a threshold here.

Reply:
as discussed, we couldn't find any timeout method on Boa API end,  thus we will handle it on our own and set 1 day (24 hour) for now

Reply:
fixed in https://github.com/CenterForOpenScience/osf.io/pull/10481/commits/040bafcf4b9cf3232e7e533e376ebd9139090f9d